### PR TITLE
Add support for parsing/stringifying fragment identifier

### LINKED ISSUE #220: Support hash
```javascript
queryString.parseUrl('https://www.omg.com/page#param1&param2');
```
Result without hash...
```javascript
{​
    query: Object {  }
​    url: "https://www.omg.com/page"
}
```

### PR DESCRIPTION
**Pull request by Mark1626**
Fixes #220 

Add a new option `parseFragment`, which will provide the hash in the `parseURL` result. The new option is to prevent it from becoming a breaking change

Seems `Fragment Identifier` is the technical jargon for it, let me know if it's ok or hash seems like a better choice

cc\ @sindresorhus

### CODE REVIEW

**Comment by sindresorhus** [comment]
The name is "fragment identifier", not "fragment". "fragment" feels a bit too generic.

Can you also add it to the readme?

**Comment by M4rk9696** [comment]
@sindresorhus Done

File: readme.md [hunk 1]
```diff
@@ -276,6 +276,19 @@ queryString.parseUrl('https://foo.bar?foo=bar');
 //=> {url: 'https://foo.bar', query: {foo: 'bar'}}
 ```
 
+#### parseFragmentIdentifier
+
+Extracts the fragment identifier from url
```
Comment (sindresorhus) [inline readme.md]: The readme and TS file should be exactly the same.

File: readme.md [hunk 2]
```diff
@@ -276,6 +276,19 @@ queryString.parseUrl('https://foo.bar?foo=bar');
...
+
+Type: `boolean`\
+Default: `false`
+
+```js
+import queryString = require('query-string');
```
Comment (sindresorhus) [inline readme.md]: Import like that is only valid TS syntax.

```suggestion
com queryString = require('query-string');
```

**Comment by sindresorhus** [comment]
Can you fix the merge conflict?

**Comment by sindresorhus** [comment]
`stringifyUrl()` was just added: https://github.com/sindresorhus/query-string/commit/16049586d5d6da4dcd42a3c8aef2b3e80ca30d69 I think it would make sense to support `fragmentIdentifier` there too so they're perfect opposites.

File: index.d.ts [hunk 3]
```diff
@@ -126,6 +141,7 @@ export function parse(query: string, options?: ParseOptions): ParsedQuery;
 export interface ParsedUrl {
 	readonly url: string;
 	readonly query: ParsedQuery;
+	readonly fragmentIdentifier?: string;
```
Comment (sindresorhus) [inline index.d.ts]: You could make `ParsedUrl` generic so `fragmentIdentifier` only exist when the `parseFragmentIdentifier` option is `true`. That would be nicer for user as it wouldn't be an optional.

**Comment by sindresorhus** [comment]
@M4rk9696 Bump

**Comment by sindresorhus** [comment]
Bump

File: index.d.ts (hunk 3 above)
Comment (M4rk9696) [reply index.d.ts:160]: @sindresorhus I didn't get get what you meant. Can you give an example

**Comment by M4rk9696** [comment]
Sorry, got busy with some work. I've made the changes except the typing. Any further changes I'll get it done ASAP and we'll get this closed soon

File: index.d.ts (hunk 3 above)
Comment (sindresorhus) [reply index.d.ts:160]: I was thinking something similar with https://github.com/sindresorhus/query-string/pull/222/files#diff-b52768974e6bc0faccb7d4b75b162c99R152, where the option value controls what's returned, but after some more thinking, I don't think it's worth complicating it.

File: index.d.ts [hunk 4]
```diff
@@ -142,6 +157,7 @@ export function parse(query: string, options?: ParseOptions): ParsedQuery;
...
```
Comment (sindresorhus) [inline index.d.ts:160]: This one is not documented in the readme.

File: readme.md [hunk 5]
```diff
@@ -383,6 +383,19 @@ The URL to stringify.
 Type: `object`
 
 Query items to add to the URL.
+#### parseFragmentIdentifier
```
Comment (sindresorhus) [inline readme.md]: This is documented in the wrong place.

**Comment by sindresorhus** [comment]
https://github.com/sindresorhus/query-string/pull/222/files#diff-0730bb7c2e8f9ea2438b52e419dd86c9R357 is no longer correct.

File: index.d.ts (hunk 4 above)
Comment (M4rk9696) [reply index.d.ts:160]: Which section would this go under?

File: readme.md [hunk 6]
```diff
@@ -350,6 +350,20 @@ queryString.parseUrl('https://foo.bar?foo=bar');
 //=> {url: 'https://foo.bar', query: {foo: 'bar'}}
 ```
 
+##### parseFragmentIdentifier
```
Comment (M4rk9696) [inline readme.md:353]: Moving this under the `parseUrl` section as this option does not apply to parse

Comment (sindresorhus) [reply readme.md:353]: You need to make it clearer that it's an option.

And https://github.com/sindresorhus/query-string/pull/222/commits/66379d5ede6e8301233096c6f85a8bcb69b0e5cf#diff-0730bb7c2e8f9ea2438b52e419dd86c9R342 is still not correct.

File: index.d.ts (hunk 4 above)
Comment (sindresorhus) [reply index.d.ts:160]: I feel like that's quite obvious: https://github.com/sindresorhus/query-string/pull/222/files#diff-0730bb7c2e8f9ea2438b52e419dd86c9R344

**Comment by Mark1626** [comment]
@sindresorhus Made the changes

File: index.d.ts [hunk 7]
```diff
@@ -142,6 +157,11 @@ export function parse(query: string, options?: ParseOptions): ParsedQuery;
 export interface ParsedUrl {
 	readonly url: string;
 	readonly query: ParsedQuery;
+
+	/**
+	The fragment identifier of the URL.
```
Comment (sindresorhus) [inline index.d.ts:162]: This needs to say when it's available. Same in the readme.

File: index.d.ts [hunk 8]
```diff
@@ -310,6 +330,27 @@ export interface StringifyOptions {
 	```
 	*/
 	readonly skipEmptyString?: boolean;
+
+	/**
+	Adds the fragment identifier to the URL.
```
Comment (sindresorhus) [inline index.d.ts:335]:
```suggestion
	Add the fragment identifier to the URL.
```

File: index.d.ts [hunk 9]
```diff
@@ -310,6 +330,27 @@ export interface StringifyOptions {
...
+	@example
+	```
+	import queryString = require('query-string');
+
+	queryString.stringifyUrl({
+		url: 'https://foo.bar',
+		query: {top: 'foo'},
+		fragmentIdentifier: 'bar'
+	}, {
+		parseFragmentIdentifier: true
```
Comment (sindresorhus) [inline index.d.ts:348]: This option name makes no sense for **stringify**Url. Should it even need an option here?

File: index.js [hunk 10]
```diff
@@ -338,22 +338,30 @@ exports.stringify = (object, options) => {
...
-	const hash = getHash(input.url);
+	let hash = getHash(input.url);
 	const query = Object.assign(parsedQueryFromUrl, input.query);
 	let queryString = exports.stringify(query, options);
 	if (queryString) {
 		queryString = `?${queryString}`;
 	}
 
+	if (options && options.parseFragmentIdentifier && input.fragmentIdentifier) {
+		hash = `#${input.fragmentIdentifier}`;
```
Comment (sindresorhus) [inline index.js:363]: Shouldn't this be properly encoded? Also needs tests for this to ensure it properly encodes.

File: readme.md [hunk 11]
```diff
@@ -350,6 +348,28 @@ queryString.parseUrl('https://foo.bar?foo=bar');
 //=> {url: 'https://foo.bar', query: {foo: 'bar'}}
 ```
 
+#### options
+
+Type: `object`
+
+The `options` are the same as for `.parse()`.
+
+Extra options are as below
```
Comment (sindresorhus) [inline readme.md:357]:
```suggestion
Extra options are as below.
```

File: index.d.ts (hunk 7 above)
Comment (Mark1626) [reply index.d.ts:162]: This has been done individually for `parseURL` and `stringifyURL` sections

File: index.d.ts [hunk 12]
```diff
@@ -166,6 +166,7 @@ export interface ParsedUrl {
 
 /**
 Extract the URL and the query string as an object.
+Optionally object contains `fragmentIdentifier` if `parseFragmentIdentifier` is `true` in options.
```
Comment (sindresorhus) [inline index.d.ts:169]:
```suggestion
If the `parseFragmentIdentifier` is `true`, the object will also contain a `fragmentIdentifier` property.
```

File: index.d.ts [hunk 13]
```diff
@@ -352,6 +356,8 @@ Stringify an object into a URL with a query string and sorting the keys. The inv
 
 Query items in the `query` property overrides queries in the `url` property.
 
+Fragment identifier in the `fragmentIdentifier` property overrides fragment identifier in the `url` property
```
Comment (sindresorhus) [inline index.d.ts:359]:
```suggestion
The `fragmentIdentifier` property overrides the fragment identifier in the `url` property.
```

File: index.d.ts (hunk 7 above)
Comment (sindresorhus) [reply index.d.ts:162]: It should be documented there too.

**Comment by sindresorhus** [comment]
https://github.com/sindresorhus/query-string/pull/222#discussion_r434604279

File: index.d.ts (hunk 7 above)
Comment (Mark1626) [reply index.d.ts]: It's not clear what you mean

File: index.d.ts [hunk 14]
```diff
@@ -159,7 +159,7 @@ export interface ParsedUrl {
 	readonly query: ParsedQuery;
 
 	/**
-	The fragment identifier of the URL.
```
Comment (sindresorhus) [inline index.d.ts]: This should stay.

File: index.d.ts [hunk 15]
```diff
@@ -159,7 +159,7 @@ export interface ParsedUrl {
...
+	If the `parseFragmentIdentifier` is `true`, the object will also contain a `fragmentIdentifier` property.
```
Comment (sindresorhus) [inline index.d.ts:162]: This text should be shortened and customized for this context.

**Comment by Mark1626** [comment]
Made the changes

File: index.d.ts (hunk 7 above)
Comment (sindresorhus) [reply index.d.ts:162]: It's good now.

**Comment by luckymore** [comment]
@Mark1626 How about this feature 😂

excute
```js
queryString.parseUrl('https://foo.bar?query=1#/hashpath?hashquery=1', { parseFragmentIdentifier: true })
```

now
```json
{
    "url": "https://foo.bar",
    "query": {
        "query": "1"
    },
    "fragmentIdentifier": "/hashpath?hashquery=1"
}
```

expect

```json
{
    "url": "https://foo.bar",
    "query": {
        "query": "1"
    },
    "params": {
        "hashquery": "1"
    },
    "fragmentIdentifier": "/hashpath?hashquery=1"
}
```

**Pull request merged**